src/intptrcast.rs

```diff
@@ -72,12 +74,45 @@ impl<'mir, 'tcx> GlobalStateInner {
                     None
                 }
             }
-        };
-        // Pointers created from integers are untagged.
-        Pointer::new(
-            alloc_id.map(|alloc_id| Tag { alloc_id, sb: SbTag::Untagged }),
-            Size::from_bytes(addr),
-        )
+        }?;
+
+        match global_state.provenance_mode {
+            ProvenanceMode::Legacy => Some(alloc_id),
+            ProvenanceMode::Permissive if global_state.exposed.contains(&alloc_id) =>
+                Some(alloc_id),
+            _ => None,
+        }
+    }
+
+    pub fn expose_addr(ecx: &MiriEvalContext<'mir, 'tcx>, alloc_id: AllocId) {
+        trace!("Exposing allocation id {:?}", alloc_id);
+
+        let mut global_state = ecx.machine.intptrcast.borrow_mut();
+        global_state.exposed.insert(alloc_id);
+    }
+
+    pub fn ptr_from_addr(ecx: &MiriEvalContext<'mir, 'tcx>, addr: u64) -> Pointer<Option<Tag>> {
+        trace!("Transmuting 0x{:x} to a pointer", addr);
+
+        // TODO: fix this at some point once we deal with function pointers
+        // Pointer::new(None, Size::from_bytes(addr))
+        Self::ptr_from_casted_addr(ecx, addr)
+    }
+
+    pub fn ptr_from_casted_addr(
+        ecx: &MiriEvalContext<'mir, 'tcx>,
+        addr: u64,
+    ) -> Pointer<Option<Tag>> {
+        trace!("Casting 0x{:x} to a pointer", addr);
+
+        let global_state = ecx.machine.intptrcast.borrow();
+
+        // Special-case NULL, it is always an invalid pointer.
+        if addr == 0 || global_state.provenance_mode == ProvenanceMode::Strict {
+            return Pointer::new(None, Size::from_bytes(addr));
+        }
+
+        Pointer::new(Some(Tag::Wildcard), Size::from_bytes(addr))
```

Comment:
This returns `Wildcard` under legacy mode, that doesn't seem right.

Reply:
As I mentioned in my response to https://github.com/rust-lang/miri/pull/2059#discussion_r875687052, the intent is to only allowing this find best allocation (there's a word for this that I can't think of right now) if the tag matches `Tag::Wildcard`, but I'm open to changes here.

Reply:
That test uses "miri-permissive-provenance" so it is unrelated to my comment.

In legacy mode, this should do the same thing as before: do ´alloc_id_from_addr` *at cast time* and use the resulting AllocId for provenance.